app-frontend/src/app/components/diagramContainer/diagramContainer.controller.js

```diff
@@ -229,6 +132,205 @@ export default class DiagramContainerController {
         });
     }
 
+    extractInputs() {
+        this.inputsJson = [];
+
+        let json = Object.assign({}, this.toolJson);
+        let inputs = [json];
+        while (inputs.length) {
+            let input = inputs.pop();
+            let args = input.args;
+            if (args) {
+                let tool = this.getToolLabel(input);
+                if (!Array.isArray(args)) {
+                    args = Object.values(args);
+                }
+                inputs = inputs.concat(args.map((a) => {
+                    return Object.assign({parent: tool}, a);
+                }));
+            } else {
+                this.inputsJson.push(input);
+            }
+        }
+    }
+
+    extractShapes() {
+
+        function shouldCreateRect(input) {
+            // We only create nodes for non-inputs or layer inputs
+            let args = input.args;
+            return args || !args && input.type === 'layer';
+        }
+
+        let nextId = 0;
+        let nodes = new Map();
+        let shapes = [];
+        let json = Object.assign({}, this.toolJson);
+        let inputs = [json];
+
+        while (inputs.length) {
+            let input = inputs.pop();
+            let rectangle;
+            let args;
+
+            // Args can be array or object, if object, convert to array
+            if (input.args) {
+                args = Array.isArray(input.args) ? input.args : Object.values(input.args);
+            } else {
+                args = [];
+            }
+
+
+            if (shouldCreateRect(input)) {
+                let rectInputs = args.filter(a => a.type === 'layer' || a.apply).length;
+                let rectOutputs = ['Output'];
+                let ports = this.createPorts(rectInputs, rectOutputs);
+                let currentId = nextId.toString();
+                let rectAttrs = {
```

Comment:
Let's move this to constructRectProps - the method signature can be changed to 
```
constructRectProps(id, input)
```
and generate the label, ports, inputs, outputs, and label inside since they aren't used anywhere else